api/envoy/service/ext_proc/v3alpha/external_processor.proto

```diff
@@ -0,0 +1,254 @@
+    // to be immediately sent to the client. It is ignored after
+    // other message types.
+    ImmediateResponse immediate_response = 7;
+  }
+
+  // Add new metadata to the Envoy context
+  google.protobuf.Struct new_metadata = 8;
+
+  // Override how parts of the HTTP request and response are processed
+  // for the duration of this particular request/response only. Servers
+  // may use this to intelligently control how requests are processed
+  // based on the headers and other metadata that they see.
+  envoy.extensions.filters.http.ext_proc.v3alpha.ProcessingMode mode_override = 9;
+
+  // Change how and if the filter waits for responses from the server
+  // for future messages. Servers may use this to switch between
+  // synchronous and asynchronous processing modes based on what they see.
+  // A message with this field may be sent even if none of the "oneof" fields
+  // have been sent, and it may be sent out-of-band with respect to
+  // the other messages if desired.
+  ProcessingModeChange mode_change = 10;
+}
+
+// The following are messages that are sent to the server.
+
+// This message is sent to the external server when the HTTP request and responses
+// are first received.
+message HttpHeaders {
+  // The HTTP request headers. If multiple headers share the same key, they
+  // must be merged according to the HTTP spec. All header keys must be
+  // lower-cased, because HTTP header keys are case-insensitive.
+  config.core.v3.HeaderMap headers = 1;
+
+  // Information about the client's connection -- downstream if this is
+  // a request and upstream if it is a response.
+  auth.v3.AttributeContext.Peer client_info = 2;
+
+  // The values of properties selected by the "request_properties"
+  // list in the configuration. Each entry in the map is the
+  // value of the specified property.
+  map<string, google.protobuf.Struct> properties = 3;
+
+  // If true, then there is no message body associated with this
+  // request or response.
+  bool no_body = 4;
+}
+
+// This message contains the message body that Envoy sends to the external server.
+message HttpBody {
+  bytes body = 1;
+
+  bool end_of_stream = 2;
+}
+
+// This message contains the trailers.
+message HttpTrailers {
+  config.core.v3.HeaderMap trailers = 1;
+}
+
+// The following are messages that may be sent back by the server.
+
+// This message must be sent in response to a RequestHeaders or ResponseHeaders
+// message.
+message HeaderResponse {
+  // Add or replace HTTP headers.
+  repeated config.core.v3.HeaderValueOption set_headers = 1;
+
+  // Remove these HTTP headers.
+  repeated string remove_headers = 2;
+}
+
+message BodyResponse {
+  // Add or replace HTTP headers.
+  repeated config.core.v3.HeaderValueOption set_headers = 1;
+
+  // Remove these HTTP headers.
+  repeated string remove_headers = 2;
+
+  // Request that the last body chunk delivered be modified.
+  BodyMutation body_mutation = 3;
```

Comment:
How does this work with streaming?  
If Envoy sends chunk A B C as they arrive, and it gets back a response I assume the modification is for chunk A not chunk C (which is the last to be delivered)

Looking at the definition it looks more like a general offset thing than anything else.

Reply:
It's right now a request / response protocol. If the sync mode is "SYNC," then Envoy sends a HttpBody message each time a "Body" is called on the filter and suspends processing until it gets a BodyResponse message. So, the BodyResponse always modifies the last body that was received by the filter, and there's no pipelining. (Although I suppose that we could add a sequence number fo each HttpBody message if you think pipelining is important, or allow pipelining in general.)

So there's one BodyMutation per HttpBody.

I could see the body_mutation field being repeated, however, but I'm not sure exactly what the use case would be.